cogboard-app/src/main/kotlin/com/cognifide/cogboard/logStorage/LogStorage.kt

```diff
@@ -0,0 +1,293 @@
+            }
+        }
+    }
+
+    /** Deletes logs not matching to the rules from the database. */
+    fun filterExistingLogs() {
+        val collection = logsCollection ?: return
+
+        val fieldName = Log.VARIABLE_DATA + "." + LogVariableData.HEADER
+        val regexes = enabledRegexes.map { regex(fieldName, it.pattern) }
+
+        if (regexes.isEmpty()) { return }
+
+        collection.deleteMany(nor(regexes))
+    }
+
+    /** Downloads new logs and inserts the to the database. Returns the number of inserted logs. */
+    private fun downloadInsertLogs(seq: Long, skipFirstLines: Long? = null): Long {
+        var sequence = seq
+        val logs = connection
+                .getLogs(skipFirstLines)
+                .mapNotNull { parserStrategy.parseLine(it) }
+                .toMutableList()
+
+        // Filter the logs by quarantine rules
+        filter(logs)
+        logs.forEach {
+            it.seq = sequence
+            sequence += 1
+        }
+
+        logsCollection?.insertMany(logs.map { it.toDocument() })
+
+        return logs.size.toLong()
+    }
+
+    /** Checks how many logs to download, downloads them and saves them to the database. */
+    private fun downloadLogs() {
+        // Get the current settings
+        val storageConfig = collectionConfiguration
+        var lastLine = storageConfig?.lastLine ?: 0
+        var seq = storageConfig?.seq ?: 0
+
+        // Get the number of lines in the file
+        val fileLineCount = connection.getNumberOfLines() ?: 0
+
+        if (fileLineCount > 0 && fileLineCount > lastLine) {
+            // Download new logs and append them
+            val inserted = downloadInsertLogs(seq, lastLine)
+            lastLine += inserted
+            seq += inserted
+        } else if (fileLineCount in 1 until lastLine) {
+            // Remove all logs and download from the beginning
+            deleteAllLogs()
+            seq = 0
+            val inserted = downloadInsertLogs(seq)
+            lastLine = inserted
+            seq += inserted
+        }
+
+        // Save the new configuration
+        saveConfiguration(LogCollectionConfiguration(config.id, lastLine, seq))
+    }
+
+    /** Returns all logs for this instance, sorted by their sequence number. */
+    private val logs: List<Log>
+    get() = logsCollection
+            ?.find()
+            ?.limit(config.logLines.toInt())
+            ?.sort(descending(Log.SEQ))
+            ?.mapNotNull { it }
+            ?.mapNotNull { Log.from(it) }
+            ?: emptyList()
+
+    /** Prepares a JSON response to be displayed */
+    private fun prepareResponse(): JsonObject {
+        return JsonObject(mapOf(
+            "variableFields" to parserStrategy.variableFields,
+            "logs" to JsonArray(logs.map { it.toJson() })
+        ))
+    }
+
+    /** Updates the logs and sends them to the widget. */
+    fun updateLogs(fetchNewLogs: Boolean) {
+        if (fetchNewLogs) {
+            // Download new logs
```

Comment:
Please remove such comments if we know exactly what is happening from the method name below.
Don't repeat yourself!

Reply:
Done.